### PR TITLE
Condition Sets can incorrectly evaluate telemetry objects that update infrequently

### PR DESCRIPTION
**Pull request by akhenry**
* Fixes https://github.com/nasa/openmct/issues/7992
* Fixes https://github.com/nasa/openmct/issues/8060

Fixes a bug in Condition Sets that sometimes caused incorrect output. There are two things that are important to know about condition sets in order to understand what's going wrong here:
1. Condition sets use short-circuit evaluation as a performance optimization, meaning that if a condition evaluates to true, subsequent conditions in the condition set will not be evaluated.
5. Criteria are only evaluated when a new telemetry value is received. Their evaluation result is cached, and the cached value used for evaluation of the condition set to which they belong. This is another performance optimization, if the telemetry value didn't change, then don't bother re-evaluating any criteria based on that telemetry object.

If you combine these together, criteria can "miss" telemetry updates because of short-circuit evaluation, and then use a stale cached value for subsequent evaluation.

### Describe your changes:

* Renames variable to avoid `condition.condition`
* Modifies condition evaluation logic [to provide a table of latest telemetry values](https://github.com/nasa/openmct/pull/8041/files#diff-2d003d796f4ffa7dc7edb76a078db1f2d8534a29ad3bce035591ef5ec73f9559R82) instead of relying upon a cached evaluation result that might be stale
* Optimization to avoid flat search of all conditions every time a value is received https://github.com/nasa/openmct/pull/8041/files#diff-486148e15c06c3403c47d65083107e8eeacf6591150e38f5690aaa6dc24c51b1L307
* Providing latest telemetry value table allows us to [retain short-circuit evaluation](https://github.com/nasa/openmct/pull/8041/files#diff-486148e15c06c3403c47d65083107e8eeacf6591150e38f5690aaa6dc24c51b1R417)
* [Fixes a bug that was preventing test data change detection from working](https://github.com/nasa/openmct/pull/8041/files#diff-486148e15c06c3403c47d65083107e8eeacf6591150e38f5690aaa6dc24c51b1R483). Test data had become reactified so the code never detected a change.

### TO DO
* [ ] Replace unit test with e2e

### All Submissions:

* [x] Have you followed the guidelines in our [Contributing document](https://github.com/nasa/openmct/blob/master/CONTRIBUTING.md)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/nasa/openmct/pulls) for the same update/change?
* [ ] Is this a [notable change](../docs/src/process/release.md) that will require a special callout in the release notes? For example, will this break compatibility with existing APIs or projects that consume these plugins?

### Author Checklist

* [x] Changes address original issue?
* [ ] Tests included and/or updated with changes?
* [x] Has this been smoke tested?
* [x] Have you associated this PR with a `type:` label? Note: this is not necessarily the same as the original issue.
* [x] Have you associated a milestone with this PR? Note: leave blank if unsure.
* [x] Testing instructions included in associated issue OR is this a dependency/testcase change?

### Reviewer Checklist

* [ ] Changes appear to address issue?
* [ ] Reviewer has tested changes by following the provided instructions?
* [ ] Changes appear not to be breaking changes?
* [ ] Appropriate automated tests included?
* [ ] Code style and in-line documentation are appropriate?

### CODE REVIEW

**Comment by akhenry** [comment]
# Testing Instructions
## Overview
* Testing is performed using a display named "7992 - Condition Set Test Display" in the Open MCT VIPER test environment.
* This display has several components:
  * "driverStatus": The parameter for driver operator status
  * "codriverStatus": The parameter for codriver operator status
  * "Condition Set Bug Test": The condition set under test. It will output values according to the truth table defined later.
  * "Driver Status and Codriver Status Plot": A plot that shows the current and historical values of driverStatus and codriverStatus. This is purely informational and is not required for the test verification. It's just provided for convenience.
 
## "Condition Set Bug Test" Truth Table
| Step | driverStatus | codriverStatus | Expected Output | Observed Output |
| ---- | ------------- | -------------- | ----------------- | ------------------ |
| 1       | NO_STATUS | NO_STATUS    | default                 | default                   |

**Review by davetsay — CHANGES_REQUESTED** [review]
Looking good. I am marking request changes, but really because you don't see the review comments unless I finalize the review.

Will smoke test and re-review when ready

File: src/plugins/condition/ConditionManager.js [hunk 1]
```diff
@@ -410,26 +400,34 @@ export default class ConditionManager extends EventEmitter {
 
     const normalizedDatum = this.createNormalizedDatum(datum, endpoint);
     const timeSystemKey = this.openmct.time.getTimeSystem().key;
```
[the remark this answers is not in the document]
Comment (akhenry) [reply src/plugins/condition/ConditionManager.js:402]: Agreed, I was following the existing implementation and I think supporting ITC from condition sets is vital, but outside the scope of this change. I'll file a followup.

File: src/plugins/condition/criterion/TelemetryCriterion.js [hunk 2]
```diff
@@ -236,7 +253,7 @@ export default class TelemetryCriterion extends EventEmitter {
       });
   }
 
-  findOperation(operation) {
+  #findOperation(operation) {
     for (let i = 0, ii = OPERATIONS.length; i < ii; i++) {
```
[the remark this answers is not in the document]
Comment (akhenry) [reply src/plugins/condition/criterion/TelemetryCriterion.js:257]: Yeah I've no idea what's going on here. Seems like maybe an attempt to cache the value of OPERATIONS.length? But why? I'll remove.

File: src/plugins/condition/ConditionManager.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (akhenry) [reply src/plugins/condition/ConditionManager.js:402]: Have filed a followup - https://github.com/nasa/openmct/issues/8063

File: src/plugins/condition/Condition.js [hunk 3]
```diff
@@ -44,48 +44,59 @@ import { getLatestTimestamp } from './utils/time.js';
...
+    if (!latestDataTable) {
       console.log('no data received');
-
       return;
     }
 
     // if all the criteria in this condition have no telemetry, we want to force the condition result to evaluate
-    if (this.hasNoTelemetry() || this.isTelemetryUsed(datum.id)) {
+    if (this.hasNoTelemetry() || this.isTelemetryUsed(telemetryIdThatChanged)) {
+      const currentTimeSystemKey = this.openmct.time.getTimeSystem().key;
```
[the remark this answers is not in the document]
Comment (akhenry) [reply src/plugins/condition/Condition.js:90]: Have filed followup https://github.com/nasa/openmct/issues/8063

**Comment by davetsay** [comment]
I'm noticing a weird edge case with Test Data. I've done the testing steps on #8060, and its fine. I then add an additional P1 test datum, and the condition ignores it. This is probably a design bug and not a bug in how Test Data works, but thought its worth mentioning.

image

**Comment by akhenry** [comment]
> I'm noticing a weird edge case with Test Data. I've done the testing steps on #8060, and its fine. I then add an additional P1 test datum, and the condition ignores it. This is probably a design bug and not a bug in how Test Data works, but thought its worth mentioning.
> 
> image

This is fine, at least for now. I'm not even sure how it _should_ behave if there are multiple test inputs defined for a single telemetry point? We should probably only allow one per telemetry point.

**Review by davetsay — APPROVED** [review]
LGTM
I made a commit directly, which just removes some commented code and a grammar fix in the test name.

**Review by davetsay — CHANGES_REQUESTED** [review]
Failing Visual Test: Needle Gauge with State Generator is now not showing the Gauge, but the State Generator
